src/index.ts

```diff
@@ -2,70 +2,105 @@ import * as core from '@actions/core';
 import GithubMetrics from './github-metrics';
 import { WebClient } from '@slack/web-api';
 import { constructSlackMessage } from './utils/slack';
-import TimeToMergeMetric from './metrics/time-to-merge';
-import MergedPRsMetric from './metrics/merged-prs';
 import { setGithubArgs } from './utils/env';
-import { getInterval, Period } from './utils/date';
 import { fetchWorkflows } from './utils/api-requests';
-import WorkflowDurationMetric from './metrics/workflow-duration';
 import debug, { enableDebugging } from './utils/debug';
+import { parse } from 'yaml';
+import {
+  MetricsConfig,
+  generateMetrics,
+} from './utils/generate-metrics-from-config';
+
+const DEFAULT_CONFIG_YML = `
+period: 'week'
+
+# Runs all metrics, by default
+metrics:
+ - name: 'pull-request/merged'
+ - name: 'pull-request/time-to-merge'
+ - name: 'workflow/duration'
+ - name: 'workflow/success'
+`.trim();
+
+// Configuration options that can be passed
+// via environment variables or, when run via
+// a github actions (GHA) workflow, GHA input values.
+type CommandLineConfiguration = {
+  githubOwner: string;
+  githubRepo: string;
+  githubToken: string;
+  slackChannelId: string;
+  slackAppToken: string;
+  postToSlack: boolean;
+  logDebugMessages: boolean;
+  metricsConfig: MetricsConfig;
+};
+
+function loadCommandLineConfiguration(): CommandLineConfiguration {
+  const githubOwner = process.env.GITHUB_OWNER || core.getInput('github-owner');
+  const githubRepo = process.env.GITHUB_REPO || core.getInput('github-repo');
+  const githubToken = process.env.GITHUB_TOKEN || core.getInput('github-token');
+  const slackChannelId =
+    process.env.SLACK_CHANNEL_ID || core.getInput('slack-channel-id');
+  const slackAppToken =
+    process.env.SLACK_APP_TOKEN || core.getInput('slack-app-token');
+  const logDebugMessages =
+    process.env.LOG_DEBUG_MESSAGES || core.getInput('log-debug-messages');
+  const postToSlack =
+    process.env.POST_TO_SLACK || core.getInput('post-to-slack');
+
+  const configYml =
+    process.env.CONFIG_YML || core.getInput('config-yml') || DEFAULT_CONFIG_YML;
+  const config = parse(configYml);
```

Comment:
What happens if yaml string is malformed? Does GH action blow up?

Reply:
Good point! It would throw an error but I'm not sure what happens when an uncaught exception occurs. I'll move the config-loading code within the `run()` try/catch as well, so it's clear that the failure is handled. 👍